src/dd/StateGeneration.cpp

```diff
@@ -0,0 +1,183 @@
+                             const std::vector<std::size_t>& nodesPerLevel,
+                             const GenerationWireStrategy strategy,
+                             Package& dd) {
+  std::random_device rd;
+  return generateRandomState(levels, nodesPerLevel, strategy, dd, rd());
+}
+
+VectorDD generateRandomState(const std::size_t levels,
+                             const std::vector<std::size_t>& nodesPerLevel,
+                             const GenerationWireStrategy strategy, Package& dd,
+                             const std::size_t seed) {
+  if (levels <= 0U) {
+    throw std::invalid_argument("Number of levels must be greater than zero");
+  }
+  if (nodesPerLevel.size() != levels - 1) {
+    throw std::invalid_argument(
+        "Number of levels - 1 must match nodesPerLevel size");
+  }
+
+  [[maybe_unused]] bool inc{};
+
+  Generator gen(seed);
+  AngleDistribution dist{0, 2. * qc::PI};
+
+  std::size_t v = levels - 1;
+  const vCachedEdge root = randomNode(static_cast<Qubit>(v), gen, dist, dd);
+
+  std::vector<vCachedEdge> prev{root};
+  for (const std::size_t n : nodesPerLevel) {
+    if (n > 2UL * prev.size()) {
+      throw std::invalid_argument(
+          "Number of nodes per level must not exceed twice the number of "
+          "nodes in the level above");
+    }
+
+    --v;
+
+    std::vector<vCachedEdge> curr(n); // Random nodes on layer v.
+    std::generate(curr.begin(), curr.end(), [&] {
+      return randomNode(static_cast<Qubit>(v), gen, dist, dd);
+    });
+
+    std::vector<std::size_t> indices(2 * prev.size()); // Indices for wireing.
+    switch (strategy) {
+    case ROUNDROBIN: {
+      std::generate(indices.begin(), indices.end(),
+                    [&n, r = 0UL]() mutable { return (r++) % n; });
+      break;
+    }
+    case RANDOM: {
+      IndexDistribution idxDist{0, n - 1};
+
+      // Make sure that each successor is connected to the previous layer.
+      auto lastOfN = indices.begin();
+      std::advance(lastOfN, n);
+      std::iota(indices.begin(), lastOfN, 0);
+
+      // Choose the rest randomly.
+      std::generate(lastOfN, indices.end(),
+                    [&idxDist, &gen]() { return idxDist(gen); });
+
+      // Shuffle to randomly interleave the resulting indices.
+      std::shuffle(indices.begin(), indices.end(), gen);
+      break;
+    }
+    }
+
+    // Wire previous layer with the current.
+    for (std::size_t i = 0; i < prev.size(); ++i) {
+      const vCachedEdge& e = prev[i];
+
+      vEdge& left = e.p->e[0];
+      vEdge& right = e.p->e[1];
+
+      // Note that the following assignments ignore the edge weights of the
+      // current layer. Since these aren't in the lookup table, we can
+      // safely do so.
+
+      left.p = curr[indices[2 * i]].p;
+      right.p = curr[indices[(2 * i) + 1]].p;
+
+      inc = dd.getUniqueTable<vNode>().incRef(left.p) &&
+            dd.getUniqueTable<vNode>().incRef(right.p);
```

Comment:
I am not 100% sure that this form of using the existing reference counting is the most efficient way to do things/saves that much.
What I had in mind was to actually bump the reference count of the respective nodes manually.

However, we should not waste too much time on this, as there is an ongoing PR that is part of unitaryHACK, which removes reference tracking altogether.
So this might not even be as important in the middle term.

Reply:
As of now, I've returned to a single `dd.incRef(...)` at the end of the function. Should we keep it that way for now, or do you think it's worth the effort to in-place increase the ref counts (given that ref counts for nodes might be history soon)?

Reply:
Yeah. That's totally fine and adequate!